Libraries/UserInterface/src/server_main_window.cpp

```diff
@@ -1,758 +1,758 @@
+	{
+		if (msg->wParam == VK_TAB)
+		{
+			HandleTabbing();
+			ret = true;
+		}
+		else if (msg->wParam == VK_RETURN)
+		{
+			OnDefaultAction();
+			ret = true;
+		}
+		else if (msg->wParam == VK_ESCAPE)
+		{
+			if (callback_)
+			{
+				if (ui_ == STREAMING)
+				{
+					callback_->DisconnectFromCurrentPeer();
+				}
+				else
+				{
+					callback_->DisconnectFromServer();
+				}
+			}
+		}
+	}
+	else if (msg->hwnd == NULL && msg->message == UI_THREAD_CALLBACK)
+	{
+		callback_->UIThreadCallback(static_cast<int>(msg->wParam),
+			reinterpret_cast<void*>(msg->lParam));
+		ret = true;
+	}
+
+	return ret;
+}
+
+void ServerMainWindow::SwitchToConnectUI()
+{
+	RTC_DCHECK(IsWindow());
+	LayoutPeerListUI(false);
+	ui_ = CONNECT_TO_SERVER;
+	LayoutConnectUI(true);
+	::SetFocus(edit1_);
+
+	if (auto_connect_)
+	{
+		::PostMessage(button_, BM_CLICK, 0, 0);
+	}
+}
+
+void ServerMainWindow::SwitchToPeerList(const std::map<int, std::string>& peers)
+{
+	LayoutConnectUI(false);
+
+	::SendMessage(listbox_, LB_RESETCONTENT, 0, 0);
+
+	AddListBoxItem(listbox_, "List of currently connected peers:", -1);
+	std::map<int,std::string>::const_iterator i = peers.begin();
+	for (; i != peers.end(); ++i)
+		AddListBoxItem(listbox_, i->second.c_str(), i->first);
+
+	ui_ = LIST_PEERS;
+	LayoutPeerListUI(true);
+	::SetFocus(listbox_);
+
+	if (auto_call_ && peers.begin() != peers.end())
+	{
+		// Get the number of items in the list
+		LRESULT count = ::SendMessage(listbox_, LB_GETCOUNT, 0, 0);
+		if (count != LB_ERR)
+		{
+			// Select the last item in the list
+			LRESULT selection = ::SendMessage(listbox_, LB_SETCURSEL, count - 1, 0);
+			if (selection != LB_ERR)
+			{
+				::PostMessage(
+					wnd_,
+					WM_COMMAND,
+					MAKEWPARAM(GetDlgCtrlID(listbox_), LBN_DBLCLK),
+					reinterpret_cast<LPARAM>(listbox_));
+			}
+		}
+	}
+}
+
+void ServerMainWindow::SwitchToStreamingUI()
+{
+	LayoutConnectUI(false);
+	LayoutPeerListUI(false);
+	ui_ = STREAMING;
+}
+
+void ServerMainWindow::MessageBox(const char* caption, const char* text, bool is_error)
+{
+	DWORD flags = MB_OK;
+	if (is_error)
+	{
+		flags |= MB_ICONERROR;
+	}
+
+	::MessageBoxA(handle(), text, caption, flags);
+}
+
+
+void ServerMainWindow::StartLocalRenderer(webrtc::VideoTrackInterface* local_video)
+{
+	local_renderer_.reset(new VideoRenderer(handle(), 1, 1, local_video));
+}
+
+void ServerMainWindow::StopLocalRenderer()
+{
+	local_renderer_.reset();
+}
+
+void ServerMainWindow::StartRemoteRenderer(webrtc::VideoTrackInterface* remote_video)
+{
+	remote_renderer_.reset(new VideoRenderer(handle(), 1, 1, remote_video));
+}
+
+void ServerMainWindow::StopRemoteRenderer()
+{
+	remote_renderer_.reset();
+}
+
+void ServerMainWindow::QueueUIThreadCallback(int msg_id, void* data)
+{
+	::PostThreadMessage(ui_thread_id_, UI_THREAD_CALLBACK,
+		static_cast<WPARAM>(msg_id), reinterpret_cast<LPARAM>(data));
+}
+
+void ServerMainWindow::SetAuthCode(const std::wstring& str)
+{
+	// TODO(bengreenier): improve implementation

```

Comment:
should we block the PR on this improvement? we could provide a similar ui to the client for this, rather than these annoying popups.